### DISCUSSION TITLE
extract returns

### DISCUSSION THREAD

**Discussion by mmann1123** [jgrss/geowombat#66] [Q&A]
Running gw.extract with polygons the returned dataframe has       

id | point | geometry | index

Just wondering how these are generated. None seems related to the polygon index. Maybe worth adding to docs?

**Comment by jgrss** [comment]
Good idea.

The `id` column is a unique numerical identifier for each polygon and is determined by the `id_column` parameter in `extract()`, `point` is a unique identifier for each point within each polygon, `geometry` is the `shapely.geometry` auto-generated by GeoPandas, and `index` must be from your polygon file.

### Using some of the built-in data

```python
from geowombat.data import l8_224078_20200518_polygons, l8_224078_20200518_B4
import geowombat as gw
import geopandas as gpd
```

### The polygons 

These will be converted to points at the native cell resolution. If `id_column` is not specified or the the column values are strings, then the `extract` function will generate a unique numerical identifier. 

```python
gpd.read_file(l8_224078_20200518_polygons).head()

        name                                           geometry
0      water  POLYGON ((737544.502 -2795232.772, 737544.502 ...
1       crop  POLYGON ((742517.658 -2798160.232, 743046.717 ...
2       tree  POLYGON ((742435.360 -2801875.403, 742458.874 ...
3  developed  POLYGON ((738903.667 -2811573.845, 738926.586 ...
```

### The extract function

The `id` column was generated from the polygon `name` column. The points are all unique, but they can be grouped by the polygon using the `id` column. The original polygon column name (`name`) is appended. I didn't specify a band name, so `1` is the raster band default.

```python
with gw.open(l8_224078_20200518_B4) as src:
    df = gw.extract(src, l8_224078_20200518_polygons)
    print(df.head())

   id  point                         geometry   name     1
0   0      0  POINT (737535.000 -2795205.000)  water  6283
1   0      1  POINT (737565.000 -2795205.000)  water  6294
2   0      2  POINT (737595.000 -2795205.000)  water  6295
3   0      3  POINT (737625.000 -2795205.000)  water  6284
4   0      4  POINT (737655.000 -2795205.000)  water  6266
```

**Reply by jgrss** [reply]
Following up on the previous comment...

The id column should be a column of integers. If `id_column` is specified and the column has strings, `extract()` should fail. If `id_column` is not specified (the default is `id_column='id'`) and the column doesn't exist, it will be created from the index values by:

```python
if id_column not in df.columns.tolist():
    df.loc[:, id_column] = df.index.values
```

If you want your own unique ids for each polygon, you could do:

```python
from geowombat.data import l8_224078_20200518_polygons
import geopandas as gpd
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df_poly = gpd.read_file(l8_224078_20200518_polygons)
df_poly['unique'] = le.fit_transform(df_poly['name'])

gw.extract(src, df_poly, id_column='unique')
```

**Comment by mmann1123** [comment]
Ok great. Thanks. Oddly the index didn't match my polygon index, odd...
I'll play around some more.

**Reply by jgrss** [reply]
Do you have an example? If you didn't provide the `id_column` value then it could be that your polygon index is not unique. ??
